packages/vite/src/node/server/index.ts

```diff
@@ -503,21 +503,28 @@ export async function createServer(
   middlewares.use(serveRawFsMiddleware(server))
   middlewares.use(serveStaticMiddleware(root, server))
 
-  // spa fallback
-  if (!middlewareMode || middlewareMode === 'html') {
-    middlewares.use(spaFallbackMiddleware(root))
-  }
-
-  // run post config hooks
-  // This is applied before the html middleware so that user middleware can
-  // serve custom content instead of index.html.
+  // We need to apply post server hooks before `spaFallbackMiddleware()`.
+  // (Because `spaFallbackMiddleware()` catches all routes.)
   postHooks.forEach((fn) => fn && fn())
 
-  if (!middlewareMode || middlewareMode === 'html') {
+  const isMiddlewareModeSSR = middlewareMode && middlewareMode !== 'html'
+
+  const isSPA = config.isSPA ?? true
```

Comment:
This still seems like a weird way to handle the default value to me. It means the default has to be specified anywhere we'd want to use `isSPA`

SvelteKit does the same thing you mentioned where we provide a framework default for some config values and allow the user to override them. It works just fine with traditionally specified config values, so I'm not sure why we'd have to do something special for this one

Reply:
`isSPA` should be resolved in [resolveConfig](https://github.com/brillout/vite/blob/5de9eb5ba32c2c3f25bbd937c35bd99aaba0fab9/packages/vite/src/node/config.ts#L292), so there is no need for this as @benmccann said.

Reply:
Done.